### PR TITLE
Add article for the WASI HTTP library

### PR DESCRIPTION
**Pull request by radu-matei**
This PR adds an article for the WASI HTTP library.

Preview: https://deploy-preview-47--deislabsio.netlify.app/posts/wasi-experimental-http/

### CODE REVIEW

File: content/posts/wasi-experimental-http.md [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+author: "@matei_radu"
+authorlink: "https://twitter.com/matei_radu"
+image: "images/twitter-card.png"
+tags: ["wasm", "wasi"]
+---
+
+Over the last year, our team has been experimenting with executing WebAssembly
+workloads on the server using the [WebAssembly System Interface][wasi] and
+[Wasmtime][wasmtime], and while more and more [proposals][proposals] get
+implemented, networking is one scenario that doesn't have a stable API yet,
```
Comment (squillace) [inline content/posts/wasi-experimental-http.md:18]: break the sentence: "on the server using WASI and wasmtime. While more and more proposalsare being implemented, networking is one scenario that doesn't have a stable API yet. This limits...."

too many phrases to parse easily.

File: content/posts/wasi-experimental-http.md [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+which limits current applications compiled to WASI, and restricts the types of
+workloads that can be executed in WASI runtimes.
+
+Today, [we are releasing a library][gh] which adds experimental and _temporary_
```
Comment (squillace) [inline content/posts/wasi-experimental-http.md:22]: "Today we are releasing" no comma needed there.

"which adds temporary support for outgoing HTTP requests for WASI modules running in Wasmtime. This is an experiment; it's intended to provide a _temporary_ workaround until until the WASI networking API is stable. We expect that once the WASI sockets proposal gets adopted and implemented in language toolchains this library will vanish."

something like that.

File: content/posts/wasi-experimental-http.md [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+support for outgoing HTTP requests for WASI modules running in Wasmtime. It is
+important to note from the beginning that this library does not aim to replace
+the WASI sockets API, but rather provide a temporary workaround until the WASI
+networking API is stable, and we expect that once [the WASI sockets
+proposal][sockets-wip] gets adopted and implemented in language toolchains, the
+need for this library would go away.
+
+### Using the new HTTP library
+
+There are two aspects to a WebAssembly library - building guest Wasm modules,
```
Comment (squillace) [inline content/posts/wasi-experimental-http.md:32]: what is a "guest module"? is it a module, or not a module? A module that isn't either? is it a NaN?

File: content/posts/wasi-experimental-http.md [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    let b = Bytes::from("sending a body in a POST request");
+    let req = req.body(Some(b)).unwrap();
+
+    let res = wasi_experimental_http::request(req).expect("cannot make request");
+    let str = std::str::from_utf8(&res.body()).unwrap().to_string();
+    // do something with the body
+}
+```
+
+From the start, the one thing to note is that we are reusing the Rust `Request`
```
Comment (squillace) [inline content/posts/wasi-experimental-http.md:57]: suggest putting your point first: "Note that because we reuse <yadda yadda>, this way of building HTTP requests should feel somewhat familiar to Rust developers and it means that a response can be sent and received [oddly the wrong way arougn? request?] using the familiar `wasi_experimental_http::request(req)`." Direct, to the point. 

Then new graph:
"Now this simple program can be compiled to .... compiles to WebAssembly. The library we are releasing also...."

File: content/posts/wasi-experimental-http.md [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+and `Response` structures from [the `http` crate][rust-http] (with `Bytes`
+request and response bodies), so this way of building HTTP requests should feel
+somewhat familiar to Rust developers. Then, a response can be sent and received
+using `wasi_experimental_http::request(req)`. This simple program can now be
+compiled to the `wasm32-wasi` target. AssemblyScript is another popular language
+that natively compiles to WebAssembly, and the library we are releasing also
+includes an AssemblyScript NPM package (with examples for how it can be used on
+[GitHub][gh]).
+
+In order to run the module in a WebAssembly runtime, we have to satisfy the new
```
Comment (squillace) [inline content/posts/wasi-experimental-http.md:67]: "To run this module in a WebAssembly runtime, we..."

File: content/posts/wasi-experimental-http.md [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+wasi.add_to_linker(&mut linker)?;
+
+// link the experimental HTTP support
+wasi_experimental_http_wasmtime::link_http(&mut linker, None)?;
+```
+
+The Wasmtime implementation also enables passing a list of allowed hosts - an
+optional and configurable list of domains or hosts that guest modules are
+allowed to send requests to. If a guest module attempts to send a request to a
+domain not explicitly allowed when the runtime was configured, it would receive
```
Comment (squillace) [inline content/posts/wasi-experimental-http.md:90]: "it receives" -- while the conditional is grammatically correct, it's hard to parse for a lot of readers so we typically use the simple present.

**Review by squillace — APPROVED** [review]
Great stuff. As per usual, I approve, but have suggested how I think it could be made to read much more smoothly. Incorporate as you see fit.

Only one thing missing, I think, is any statement about how this should be seen with respect to  WASI spec work. We say it's only temporary, and I'm good with that. Are we going to say that this will specifically only be available here, or can anyone use it with wasi work anywhere, or....

....I'm expecting something like a general statement of who and where this **could** be used and if they do, what special compiling you might have to do. Does that make sense?

Basically, the suggestion is to write it such that a complete or mostly new person would have the basic questions answered even though the community would not need that extra information.

**Comment by radu-matei** [comment]
I addressed all points, @squillace.
Thank you!

**Comment by squillace** [comment]
my pleasure, it's a great thing!

**Comment by radu-matei** [comment]
I realized I didn't address "the one thing missing", re: spec work.

This is only available if the person starting the runtime explicitly adds the library.
There is no plan to make this an "official" WASI API, because a) it's not really a new WASI API, it's separate, and b) this would be superseded by the sockets work.

Hence the temporary nature of the project.
Would including some form of that work?

**Comment by squillace** [comment]
yeah, totes!

**Review by thomastaylor312 — APPROVED** [review]
I like this!
